packages/puppeteer-core/src/api/locators/Locator.ts

```diff
@@ -162,40 +234,449 @@ export abstract class Locator<T> extends EventEmitter {
+    signal?: AbortSignal
+  ): Observable<never> => {
+    if (!this.#waitForEnabled) {
+      return EMPTY;
+    }
+    return from(
+      handle.frame.page().waitForFunction(
+        element => {
+          if ('disabled' in element && typeof element.disabled === 'boolean') {
+            return !element.disabled;
+          }
+          return true;
+        },
+        {
+          timeout: this.timeout,
+          signal,
+        },
+        handle
+      )
+    ).pipe(ignoreElements());
+  };
+
+  /**
+   * Compares the bounding box of the element for two consecutive animation
+   * frames and waits till they are the same.
+   */
+  #waitForStableBoundingBoxIfNeeded = <ElementType extends Element>(
+    handle: HandleFor<ElementType>
+  ): Observable<never> => {
+    if (!this.#waitForStableBoundingBox) {
+      return EMPTY;
+    }
+    return defer(() => {
+      // Note we don't use waitForFunction because that relies on RAF.
+      return from(
+        handle.evaluate(element => {
+          return new Promise<[BoundingBox, BoundingBox]>(resolve => {
+            window.requestAnimationFrame(() => {
+              const rect1 = element.getBoundingClientRect();
+              window.requestAnimationFrame(() => {
+                const rect2 = element.getBoundingClientRect();
+                resolve([
+                  {
+                    x: rect1.x,
+                    y: rect1.y,
+                    width: rect1.width,
+                    height: rect1.height,
+                  },
+                  {
+                    x: rect2.x,
+                    y: rect2.y,
+                    width: rect2.width,
+                    height: rect2.height,
+                  },
+                ]);
+              });
+            });
+          });
+        })
+      );
+    }).pipe(
+      first(([rect1, rect2]) => {
+        return (
+          rect1.x === rect2.x &&
+          rect1.y === rect2.y &&
+          rect1.width === rect2.width &&
+          rect1.height === rect2.height
+        );
+      }),
+      retry({delay: RETRY_DELAY}),
+      ignoreElements()
+    );
+  };
+
+  /**
+   * Checks if the element is in the viewport and auto-scrolls it if it is not.
+   */
+  #ensureElementIsInTheViewportIfNeeded = <ElementType extends Element>(
+    handle: HandleFor<ElementType>
+  ): Observable<never> => {
+    if (!this.#ensureElementIsInTheViewport) {
+      return EMPTY;
+    }
+    return from(handle.isIntersectingViewport({threshold: 0})).pipe(
+      filter(isIntersectingViewport => {
+        return !isIntersectingViewport;
+      }),
+      mergeMap(() => {
+        return from(handle.scrollIntoView());
+      }),
+      mergeMap(() => {
+        return defer(() => {
+          return from(handle.isIntersectingViewport({threshold: 0}));
+        }).pipe(first(identity), retry({delay: RETRY_DELAY}), ignoreElements());
+      })
+    );
+  };
+
+  /**
+   * @internal
+   */
+  clickImpl<ElementType extends Element>(
```

Comment:
Not a fan of the `Impl` surfix, a `_` prefix sound to be more consistent with internal use.